### ISSUE TITLE
TracerBoolConversionError since last release

### ISSUE THREAD

**Issue by Lilly-May** [google/jax#18979]
### Description

Hi!
We've encountered an issue in our project that utilizes JAX for various computations. Since the latest release, we've been encountering a `TracerBoolConversionError`. As far as we understand, this error is not directly thrown by our code but seems to be originating from JAX internally.
We were unable to generate a minimal code example to reproduce the error, so unfortunately, we can only provide you with the relevant section of our code that appears to trigger the error (to run it, you would have to install pertpy via `pip install pertpy` first):

```
import pertpy as pt

# load data and create model
haber_cells = pt.dt.haber_2017_regions()
sccoda = pt.tl.Sccoda()
mdata = sccoda.load(haber_cells, type="cell_level", generate_sample_level=True, cell_type_identifier="cell_label", \
                    sample_identifier="batch", covariate_obs=["condition"])
mdata = sccoda.prepare(mdata, formula="condition", reference_cell_type="Endocrine")

# this method throws the error
sccoda.run_nuts(mdata, num_warmup=100, num_samples=1000, rng_key=42)
```

In the following, I've included the last part of the Error Traceback. Since the entire Traceback is quite long, I've posted it in [this issue](https://github.com/theislab/pertpy/issues/454) and not here.

```
File /opt/homebrew/Caskroom/mambaforge/base/envs/pertpy1/lib/python3.10/site-packages/jax/_src/numpy/lax_numpy.py:4960, in _preprocess_slice.<locals>.clamp_index(i, which)
   4956 except core.InconclusiveDimensionOperation as e:
   4957   raise core.InconclusiveDimensionOperation(
   4958       f"In slice with non-constant elements the {which} ({i}) must " +
   4959       f"be resolved statically if it is >= 0.\nDetails: {e}")
-> 4960 if i_ge_0:
   4961   if step_gt_0:
   4962     # min(i, axis_size)
   4963     return axis_size - core.non_negative_dim(axis_size - i)

    [... skipping hidden 1 frame]

File /opt/homebrew/Caskroom/mambaforge/base/envs/pertpy1/lib/python3.10/site-packages/jax/_src/core.py:1468, in concretization_function_error.<locals>.error(self, arg)
   1467 def error(self, arg):
-> 1468   raise TracerBoolConversionError(arg)

TracerBoolConversionError: Attempted boolean conversion of traced array with shape bool[]..
The error occurred while tracing the function _body_fn at /opt/homebrew/Caskroom/mambaforge/base/envs/pertpy1/lib/python3.10/site-packages/numpyro/infer/hmc_util.py:1001 for while_loop. This value became a tracer due to JAX operations on these lines:

  operation a:bool[] = ge b c
    from line /Users/mamba/Documents/Code/pertpy/pertpy/tools/_coda/_sccoda.py:282 (model)
See https://jax.readthedocs.io/en/latest/errors.html#jax.errors.TracerBoolConversionError
```

Everything runs smoothly with JAX 0.4.20, so the issue seems to stem from the new release. Unfortunately, I couldn't identify any new features in the latest release that would cause this behavior.

Thanks in advance for any assistance!

### What jax/jaxlib version are you using?

jax 0.4.21, jaxlib 0.4.21

### Which accelerator(s) are you using?

CPU

### Additional system info?

Python 3.10.12 (but code also fails for Python 3.9 and 3.11); numpy 1.24.1.  Tested on both Linux and MacOS; the bug persists in both cases.

**Comment by jakevdp** [comment]
Thanks for the report, and sorry for the breakage. I suspect this is due to #18679 or #18818

@gnecula, could you take a look?

**Comment by gnecula** [comment]
I am trying to reproduce, but I get the following error:
```
  File "/Users/necula/Source/repro_18979/main.py", line 5, in <module>
    sccoda = pt.tl.Sccoda()
AttributeError: module 'pertpy.tools' has no attribute 'Sccoda'. Did you mean: '_coda'?
```

I will try to create my own repro guessing where the problem is, but if you can help fix the repro it may be faster.

**Comment by gnecula** [comment]
I was able to reproduce your problem, but only with JAX before #18818, which fixed an error with the same error message you are reporting. #18818 landed on December 5th, which is a bit after we released 0.4.21, which you are using. Please try to upgrade to the latest JAX release and let us know if you still see the error.

**Comment by Lilly-May** [comment]
Hi, thank you so much for looking into this! Indeed, everything works fine when updating JAX. Thank you @gnecula!
